### ISSUE TITLE
Implement game object GUIDs

### ISSUE THREAD

**Issue by ghost** [adventuregamestudio/ags#483]
It is desirable to add real unique identifiers to AGS game objects, _to be used in the program code_, **at least on the editor's side**, because it is almost impossible to keep ID references to these with 100% reliability. At the moment objects may be addressed by the numeric index in the list (breaks if an object is removed from the middle) or script name (but it is not present in some entities, and for some reason not mandatory in others).

Script name might be a good candidate, but it's not always mandatory, and may be freely edited by the user. If the first issue can be fixed, and you think the second is not a big deal, then script name may be used as global UID.

On the other hand, it could be a new GUID field, automatically generated and assigned by the Editor, that cannot be changed (unless user hacks into the game file, but in that case it's their own fault).

NOTE: this task will definitely interfere with room format, so it might be wanted to wait until room's open format (or maybe not).

**Comment by ivan-mogilko** [comment]
tbh I don't clearly remember what problems I implied when writing this ticket. It could've been some issues with the editor where object references had to match, or infamous problem of game saves getting broken by adding an object.

But I've been given this a small thought recently, and I believe this has to be reconsidered through once again at a later stage, having more broad engine system in mind.

For example, if speaking about game saves:
* Imagine in the future we will create all game objects dynamically at runtime, and there will be no more concept of a "static" object created in editor that persists infinitely and must be present in game at all times. In such case, objects restored from the save do not even have to match ones created at design time, which renders autogenerated UIDs mostly useless, at least for that particular case.
* On another hand, if we expect to match preexisting objects with ones from save, ScriptName might actually be a more valid field to use. The reason is, if UID is strictly autogenerated hidden field, then deleting an object in editor by mistake could lead to breaking all existing saves, as even if this object gets recreated it will have different UID, which user won't likely notice. At least mandatory ScriptName is easier to notice, fix and maintain perhaps (it's also may be generated by default if user did not want to add one themself).

In other words, it's best to first clearly define a system of game objects we want to have, then it will become more clear whether we need a UID, what it's purpose and how it's created. Thus there should not be a ticket like this without a connection to a bigger plan.
